child-processes/index.md

```diff
@@ -1,8 +1,189 @@
+
+```javascript
+
+const fibonacci = (num) => num <= 1 ? 1 : fibonacci(num - 1) + fibonacci(num - 2)
+
+process.on('message', ({ n }) => {
+  process.send({ fib: fibonacci(n), n })
+  // optional - there is no reason why this child process
+  // can't be called multiple times.
+  process.exit()
+})
+ 
+```
+
+The parent process creates 3 child processes, and passes a range of numbers to them for calculating.
+
+```javascript
+const { fork } = require('child_process')
+
+const child1 = fork('fork-child')
+const child2 = fork('fork-child')
+const child3 = fork('fork-child')
+
+// send data to the child process to perform the calculation
+child1.send({ n: 5 });
+child2.send({ n: 10 });
+child3.send({ n: 45 });
+
+child1.on('message', (m) => {
+  console.log('PARENT child1 got message:', m);
+});
+child2.on('message', (m) => {
+  console.log('PARENT child2 got message:', m);
+});
+child3.on('message', (m) => {
+  console.log('PARENT child3 got message:', m);
+});
+
+child1.on('exit', () => {
+  console.log('child1 exited');
+});
+child2.on('exit', () => {
+  console.log('child2 exited');
+});
+child3.on('exit', () => {
+  console.log('child3 exited');
+});
+```
+
+## ChildProcess Events
+
+The `child` processes communicates by emitting events to let the `parent` know what is going on.
+
+| Event Name   | Reason For Emitting                                                         |
+|--------------|-----------------------------------------------------------------------------|
+| `disconnect` | The parent process manually calls `child.disconnect`                        |
+| `error`      | The process could not be spawned or killed.                                 |
+| `exit`       | The exit code for the child and the optional signalthat was used to terminate it. When null, imples the child process exited normally. |
+| `close`      | The `stdio` streams of a child process get closed.                          |
+| `message`    | This `child` process uses the `send` method to communicate with the parent. |
+
+## A note on options
+
+These processes all accept an optional options object that allow us to control the context that the processes are run within. These vary for each method, and are described in detail in the [node documentation for `child_process`](https://nodejs.org/api/child_process.html).
+
+## Summary
+
+Parents `spawm`, `fork` or `exec` child processes, and communicate via events or pipes.
+
+Take me to [cheat sheet]({{ "/cheatsheet/#child-process" | url }})
+
+## Exercise
+
+Create a child process for doing some manipulation of a file or URL, and build a parent process that controls a number of these processes in parallel.
```

Comment:
Should this have a hidden suggested-solution? If too much to do, we can leave it for now.

Reply:
Yeah - I think there is more value in populating the other bits of content and we can come back to this, or maybe I can put a note in for someone to do one and put it up as a PR :)

Reply:
Added comment - thanks for all the comments and keeping me honest. Made a big improvement to this piece.